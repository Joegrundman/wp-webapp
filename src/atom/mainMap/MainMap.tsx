import * as React from 'react';

interface IMainMapProps {
  url: string;
  getMapContext: (ctx: CanvasRenderingContext2D | null) => void;
}

interface IMainMapState {
  height: number;
  width: number;
}

class MainMap extends React.PureComponent<IMainMapProps, IMainMapState> {
  public canvas: HTMLCanvasElement | null;
  public ctx: CanvasRenderingContext2D | null;
  public image: HTMLImageElement;

  constructor(props: IMainMapProps) {
    super(props);
    this.state = {
      height: 0,
      width: 0,
    }

    this.canvas = null;
    this.ctx = null;
    this.image = new Image();
  }

  public componentDidMount(): void {
    this.canvas = document.getElementById('mapCanvas') as HTMLCanvasElement;
    if(this.canvas) {
      this.ctx = this.canvas.getContext('2d') as CanvasRenderingContext2D;
      this.updateMap();
    }
  }

  public componentDidUpdate() {
    this.updateMap();
  }

  public updateMap(): void {
    const { ctx, image } = this;
    if(!ctx) {
      return
    }
    this.image.onload = () => {
      this.setState({
        height: this.image.height,
        width: this.image.width
      },
      () => {
        ctx.drawImage(image, 0, 0, image.width, image.height)
        if(this.ctx) {
          this.props.getMapContext(this.ctx);
        }
      })
      

    }

    image.src = this.props.url;
    image.crossOrigin = 'anonymous';    
  }

  public render(): JSX.Element {
    const { height, width } = this.state;
    return (
      <canvas id="mapCanvas" height={height} width={width}>Your browser does not support this application</canvas>
    );
  }
}

export default MainMap;