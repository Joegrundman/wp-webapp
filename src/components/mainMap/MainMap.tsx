import * as React from 'react'

export interface IMapOpts {
  rect: DOMRect
}
export interface IMainMapProps {
  url: string
  getMapContext: (mapCanvas: HTMLCanvasElement, mapCtx: CanvasRenderingContext2D | null, opts: any) => void
}

export interface IMainMapState {
  height: number
  width: number
}

class MainMap extends React.PureComponent<IMainMapProps, IMainMapState> {
  public backgroundCanvas: HTMLCanvasElement
  public backgroundCtx: CanvasRenderingContext2D
  public mapCanvas: HTMLCanvasElement 
  public mapCtx: CanvasRenderingContext2D
  public rect: DOMRect
  public image: HTMLImageElement
  public mapCanvasRef: React.RefObject<HTMLCanvasElement>
  public backgroundCanvasRef: React.RefObject<HTMLCanvasElement>

  constructor(props: IMainMapProps) {
    super(props)
    this.state = {
      height: 0,
      width: 0,
    }
  
    this.image = new Image()
    this.mapCanvasRef = React.createRef()
    this.backgroundCanvasRef = React.createRef()
  }

  public componentDidMount(): void {
    this.mapCanvas = document.getElementById('mapCanvas') as HTMLCanvasElement;
    this.rect = this.mapCanvas.getBoundingClientRect() as DOMRect;
    this.backgroundCanvas = document.getElementById('backgroundCanvas') as HTMLCanvasElement;
    this.mapCtx = this.mapCanvas.getContext('2d') as CanvasRenderingContext2D;
    this.backgroundCtx = this.backgroundCanvas.getContext('2d') as CanvasRenderingContext2D;
    this.updateMap()
  }

  public componentDidUpdate(props: IMainMapProps): void {
    if(this.props.url !== props.url) {
      this.updateMap()
    }
  }

  public onMapLoad = (): void => {
    this.setState({
      height: this.image.height,
      width: this.image.width
    },
    (): void => {

      this.backgroundCtx.drawImage(this.image, 0, 0, this.image.width, this.image.height)

      const opts: IMapOpts = {
        rect: this.rect
      }

      this.props.getMapContext(
        this.mapCanvas as HTMLCanvasElement,
        this.mapCtx as CanvasRenderingContext2D,
        opts
      )
    })
  }

  public updateMap(): void {
    this.image.crossOrigin = 'Anonymous'
    this.image.src = this.props.url
    this.image.onload = this.onMapLoad
  }

  public render(): JSX.Element {
    const { height, width }: { height: number, width: number} = this.state

    const canvasPosition: any = {
      left: 0,
      position: 'absolute',  
      top: 0
    }
    return (
      <div style={{position :"relative" }}>    
        <canvas
          id="backgroundCanvas"
          style={canvasPosition}
          ref={this.backgroundCanvasRef}
          height={height}
          width={width}
        />
        <canvas
          id="mapCanvas"
          style={canvasPosition}
          ref={this.mapCanvasRef}
          height={height}
          width={width}
        >Your browser does not support this application
        </canvas>
      </div>
    );
  }
}

export default MainMap