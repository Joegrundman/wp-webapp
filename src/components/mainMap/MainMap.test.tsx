import { shallow, ShallowWrapper } from 'enzyme'
import * as React from 'react'
import MainMap, { IMainMapProps } from './MainMap'

describe('<MainMap/>', () => {

  jest.spyOn(document, 'getElementById').mockImplementation(() => {
    return {
      getBoundingClientRect: jest.fn(() => ({x: 20, y: 30})),
      getContext: jest.fn(() => ({
        drawImage: jest.fn()
      }))
    }
  })

  const defaultProps = {
    getMapContext: jest.fn(),
    url: 'eur',  
  }
  const shallowComponent = (props: IMainMapProps): ShallowWrapper =>
    shallow(<MainMap {...props}/>)

  describe('@render', () => {
    it('should render as expected with defaultProps', () => {
      const component = shallowComponent(defaultProps)
      expect(component).toMatchSnapshot();
    })

    it('should render as expected with exact height and width in state', () => {
      const component = shallowComponent(defaultProps)
      component.setState({
        height: 300,
        width: 500
      })
      expect(component).toMatchSnapshot();
    })
  })

  describe('@lifecycle', () => {
    describe('componentDidUpdate', () => {
      it('should call updateMap if the passed in url differs from the current url', () => {
        const component: ShallowWrapper = shallowComponent(defaultProps)
        const instance = component.instance() as MainMap
        instance.updateMap = jest.fn()
        instance.componentDidUpdate({ ...defaultProps, url: 'pac' })
        expect(instance.updateMap).toHaveBeenCalled()
      })

      it('should not call updateMap if the passed in url is the same as current url', () => {
        const component: ShallowWrapper = shallowComponent(defaultProps)
        const instance = component.instance() as MainMap
        instance.updateMap = jest.fn()
        instance.componentDidUpdate({ ...defaultProps })
        expect(instance.updateMap).not.toHaveBeenCalled()
      })
    })
  })
})
