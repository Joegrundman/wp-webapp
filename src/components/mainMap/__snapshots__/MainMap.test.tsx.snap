// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainMap/> @render should render as expected with defaultProps 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainMap
    getMapContext={[MockFunction]}
    url="eur"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <canvas
          height={0}
          id="backgroundCanvas"
          style={
            Object {
              "left": 0,
              "position": "absolute",
              "top": 0,
            }
          }
          width={0}
        />,
        <canvas
          height={0}
          id="mapCanvas"
          style={
            Object {
              "left": 0,
              "position": "absolute",
              "top": 0,
            }
          }
          width={0}
        >
          Your browser does not support this application
        </canvas>,
      ],
      "style": Object {
        "position": "relative",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "height": 0,
          "id": "backgroundCanvas",
          "style": Object {
            "left": 0,
            "position": "absolute",
            "top": 0,
          },
          "width": 0,
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "canvas",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Your browser does not support this application",
          "height": 0,
          "id": "mapCanvas",
          "style": Object {
            "left": 0,
            "position": "absolute",
            "top": 0,
          },
          "width": 0,
        },
        "ref": Object {
          "current": null,
        },
        "rendered": "Your browser does not support this application",
        "type": "canvas",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <canvas
            height={0}
            id="backgroundCanvas"
            style={
              Object {
                "left": 0,
                "position": "absolute",
                "top": 0,
              }
            }
            width={0}
          />,
          <canvas
            height={0}
            id="mapCanvas"
            style={
              Object {
                "left": 0,
                "position": "absolute",
                "top": 0,
              }
            }
            width={0}
          >
            Your browser does not support this application
          </canvas>,
        ],
        "style": Object {
          "position": "relative",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "height": 0,
            "id": "backgroundCanvas",
            "style": Object {
              "left": 0,
              "position": "absolute",
              "top": 0,
            },
            "width": 0,
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "canvas",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Your browser does not support this application",
            "height": 0,
            "id": "mapCanvas",
            "style": Object {
              "left": 0,
              "position": "absolute",
              "top": 0,
            },
            "width": 0,
          },
          "ref": Object {
            "current": null,
          },
          "rendered": "Your browser does not support this application",
          "type": "canvas",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<MainMap/> @render should render as expected with exact height and width in state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainMap
    getMapContext={[MockFunction]}
    url="eur"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <canvas
          height={300}
          id="backgroundCanvas"
          style={
            Object {
              "left": 0,
              "position": "absolute",
              "top": 0,
            }
          }
          width={500}
        />,
        <canvas
          height={300}
          id="mapCanvas"
          style={
            Object {
              "left": 0,
              "position": "absolute",
              "top": 0,
            }
          }
          width={500}
        >
          Your browser does not support this application
        </canvas>,
      ],
      "style": Object {
        "position": "relative",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "height": 300,
          "id": "backgroundCanvas",
          "style": Object {
            "left": 0,
            "position": "absolute",
            "top": 0,
          },
          "width": 500,
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "canvas",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Your browser does not support this application",
          "height": 300,
          "id": "mapCanvas",
          "style": Object {
            "left": 0,
            "position": "absolute",
            "top": 0,
          },
          "width": 500,
        },
        "ref": Object {
          "current": null,
        },
        "rendered": "Your browser does not support this application",
        "type": "canvas",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <canvas
            height={300}
            id="backgroundCanvas"
            style={
              Object {
                "left": 0,
                "position": "absolute",
                "top": 0,
              }
            }
            width={500}
          />,
          <canvas
            height={300}
            id="mapCanvas"
            style={
              Object {
                "left": 0,
                "position": "absolute",
                "top": 0,
              }
            }
            width={500}
          >
            Your browser does not support this application
          </canvas>,
        ],
        "style": Object {
          "position": "relative",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "height": 300,
            "id": "backgroundCanvas",
            "style": Object {
              "left": 0,
              "position": "absolute",
              "top": 0,
            },
            "width": 500,
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "canvas",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Your browser does not support this application",
            "height": 300,
            "id": "mapCanvas",
            "style": Object {
              "left": 0,
              "position": "absolute",
              "top": 0,
            },
            "width": 500,
          },
          "ref": Object {
            "current": null,
          },
          "rendered": "Your browser does not support this application",
          "type": "canvas",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
